syz-verifier/verifier.go

```diff
@@ -31,24 +36,180 @@ type Verifier struct {
 	// - <workdir>/corpus.db: corpus with interesting programs
 	// - <workdir>/<OS-Arch>/instance-x: per VM instance temporary files
 	// grouped by OS/Arch
-	workdir       string
-	crashdir      string
-	resultsdir    string
-	target        *prog.Target
-	runnerBin     string
-	executorBin   string
-	choiceTable   *prog.ChoiceTable
-	rnd           *rand.Rand
-	progIdx       int
-	addr          string
-	srv           *RPCServer
-	calls         map[*prog.Syscall]bool
-	reasons       map[*prog.Syscall]string
-	reportReasons bool
-	stats         *Stats
-	statsWrite    io.Writer
-	newEnv        bool
-	reruns        int
+	workdir           string
+	crashdir          string
+	resultsdir        string
+	target            *prog.Target
+	runnerBin         string
+	executorBin       string
+	progGeneratorInit sync.WaitGroup
+	choiceTable       *prog.ChoiceTable
+	rnd               *rand.Rand
+	progIdx           int
+	addr              string
+	srv               *RPCServer
+	calls             map[*prog.Syscall]bool
+	reasons           map[*prog.Syscall]string
+	reportReasons     bool
+	stats             *Stats
+	statsWrite        io.Writer
+	newEnv            bool
+	reruns            int
+
+	// We use single queue for every kernel environment.
+	tasksMutex     sync.Mutex
+	onTaskAdded    *sync.Cond
+	kernelEnvTasks [][]*ExecTaskQueue
+}
+
+func (vrf *Verifier) Init() {
+	vrf.progGeneratorInit.Add(1)
+
+	vrf.onTaskAdded = sync.NewCond(&vrf.tasksMutex)
+
+	vrf.kernelEnvTasks = make([][]*ExecTaskQueue, len(vrf.pools))
+	for i := range vrf.kernelEnvTasks {
+		vrf.kernelEnvTasks[i] = make([]*ExecTaskQueue, EnvironmentsCount)
+	}
+
+	for i := range vrf.kernelEnvTasks {
+		for j := range vrf.kernelEnvTasks[i] {
+			vrf.kernelEnvTasks[i][j] = MakeExecTaskQueue()
+		}
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+	vrf.srv = srv
+}
+
+func (vrf *Verifier) StartProgramsAnalysis() {
+	go func() {
+		vrf.progGeneratorInit.Wait()
+
+		type AnalysisResult struct {
+			Diff []*ExecResult
+			Prog *prog.Prog
+		}
+
+		results := make(chan *AnalysisResult)
+		go func() {
+			for result := range results {
+				if result.Diff != nil {
+					vrf.SaveDiffResults(result.Diff, result.Prog)
+				}
+			}
+		}()
+
+		sem := semaphore.NewWeighted(int64(100))
```

Comment:
This looks simpler to me:
```
for i := 0; i < 100; i++ {
  go func() {
    for {
      p := vrf.generate()
      results <- &AnalysisResult{vrf.TestProgram(p), p}
    }
  }()
}
```

And if `TestProgram` returns something self-contained, then this can be:
```
for i := 0; i < 100; i++ {
  go func() {
    for {
      results <- vrf.TestProgram(vrf.generate())
    }
  }()
}
```

Reply:
My intention was to have some amount (here 100) programs under analysis to be sure that VMs will not wait for the tasks.
Current strange construction assumes that vrf.generate() is not thread safe and we want to have at most 100 parallel vrf.TestProgram() calls.

Reply:
Fixed. Thanks.